linux/flatpak/flatpak-embed-crates.sh

```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
+#
+# Embed a config.toml into a tarball which uses vendored crates from the
+# flatpak-cargo-generator script.
+
+# Create a temporary directory for the archive
+TMPDIR=$(mktemp -d)
+function cleanup(){
+  rm -rf $TMPDIR
+}
+trap cleanup EXIT
+
+# Decompress the tarball
+case $1 in
+  *.tar.gz)
+    gunzip -c $1 > $TMPDIR/archive.tar
+    ;;
+  *)
+    echo "Unsupported archive: $1" >&2
+    exit 1
+    ;;
+esac
+
+# Embed the config file into the tarball
+ROOTDIR=$(tar tf $TMPDIR/archive.tar | head -1 | cut -f1 -d/)
+mkdir -p $TMPDIR/$ROOTDIR/.cargo
+cat << EOF > $TMPDIR/$ROOTDIR/.cargo/config.toml
+[source.vendored-sources]
+directory = "${PWD}/cargo/vendor"
+
+[source.crates-io]
+replace-with = "vendored-sources"
+EOF
+tar -rf $TMPDIR/archive.tar -C $TMPDIR $ROOTDIR
+
+# Recompress the tarball
+case $1 in
+  *.tar.gz)
+    gzip -c $TMPDIR/archive.tar > $1
```

Comment:
should we use `-9` to improve compression?

Reply:
There is a little subtlety going on here, but we are making use of the `-r` or `--append` option to `tar` here in an attempt to minimize the possible changes to the pip package's tarball. This simply appends a record to the archive to insert a single file without touching anything else.

The downside is that this option only works on uncompressed archives, which means we have to decompress the archive first (into a plain `.tar` file) even though we don't need to extract the archive. Which is what these case statements are doing.

Reply:
The modification to this tarball is only done locally to the flatpak build directory, and the modified archive is never uploaded or stored anywhere. Therefore there isn't really any benefit to a more aggressive compression level.

Reply:
Awesome, thanks for explaining 😃